python/ray/_raylet.pyx

```diff
@@ -263,73 +275,131 @@ class StreamingObjectRefGenerator:
-                        if self._generator_task_completed_time is None:
-                            self._generator_task_completed_time = time.time()
-
-            # Currently, since the ordering of intermediate result report
-            # is not guaranteed, it is possible that althoug the task
-            # has succeeded, all of the object references are not reported
-            # (e.g., when there are network failures).
-            # If all the object refs are not reported to the generator
-            # within 30 seconds, we consider is as an unreconverable error.
-            if self._generator_task_completed_time:
-                if (time.time() - self._generator_task_completed_time
-                        > unexpected_network_failure_timeout_s):
-                    # It means the next wasn't reported although the task
-                    # has been terminated 30 seconds ago.
-                    self._generator_task_exception = AssertionError
-                    assert False, "Unexpected network failure occured."
-
-            if timeout_s != -1 and time.time() - last_time > timeout_s:
-                return ObjectRef.nil()
-
-            # 100us busy waiting
+            error_ref = self._handle_error(
+                last_time,
+                timeout_s,
+                unexpected_network_failure_timeout_s)
+            if error_ref is not None:
+                return error_ref
+
             time.sleep(sleep_interval_s)
-            obj = self._handle_next()
+            obj = self._handle_next_sync()
+
+        return obj
+
+    async def _next_async(
+            self,
+            timeout_s: float = -1,
+            sleep_interval_s: float = 0.0001,
+            unexpected_network_failure_timeout_s: float = 30):
+        """Same API as _next_sync, but it is for async context."""
+        obj = await self._handle_next_async()
+        last_time = time.time()
+
+        # The generator ref will be None if the task succeeds.
+        # It will contain an exception if the task fails by
+        # a system error.
+        while obj.is_nil():
+            error_ref = self._handle_error(
+                last_time,
+                timeout_s,
+                unexpected_network_failure_timeout_s)
+            if error_ref is not None:
+                return error_ref
+
+            await asyncio.sleep(sleep_interval_s)
+            obj = await self._handle_next_async()
+
         return obj
 
-    def _handle_next(self) -> ObjectRef:
+    async def _handle_next_async(self):
         try:
-            if hasattr(self.worker, "core_worker"):
-                obj = self.worker.core_worker.try_read_next_object_ref_stream(
-                    self._generator_ref)
-                return obj
-            else:
-                raise ValueError(
-                    "Cannot access the core worker. "
-                    "Did you already shutdown Ray via ray.shutdown()?")
+            return self._handle_next()
```

Comment:
hmm, how comes this become async function? should add inline comments

Reply:
My understanding is that the difference between sync and async is `time.sleep()` vs `asyncio.sleep()`?

Reply:
Also the function signature & we use await ref instead of ray.get